README
NAME
    Data::UUID::Concise - Encode UUIDs to be more concise or communicable

VERSION
    version 0.121082

SYNOPSIS
        use Data::UUID::Concise;

        my $duc = Data::UUID::Concise->new();
        my $encoded_uuid = $duc->encode((Data::UUID->new)->create);
        my $decoded_uuid = $duc->decode('M55djt9tt4WoFaL68da9Ef');

ATTRIBUTES
  alphabet
    This is the collection of symbols that are used for the encoding scheme.
    By default, a reasonably unambiguous set of characters is used that is
    reminiscent of the base 58 scheme used by a rather prominent photo
    site's URL shortener.

METHODS
  encode
    Encode a Data::UUID instance as a string with the appropriate set of
    symbols.

  decode
    Decode a string with the appropriate set of symbols and return a
    Data::UUID instance representing the decoded UUID.

SUPPORT
  Perldoc
    You can find documentation for this module with the perldoc command.

      perldoc Data::UUID::Concise

  Websites
    The following websites have more information about this module, and may
    be of help to you. As always, in addition to those websites please use
    your favorite search engine to discover more resources.

    *   Search CPAN

        The default CPAN search engine, useful to view POD in HTML format.

        <http://search.cpan.org/dist/Data-UUID-Concise>

    *   CPAN Ratings

        The CPAN Ratings is a website that allows community ratings and
        reviews of Perl modules.

        <http://cpanratings.perl.org/d/Data-UUID-Concise>

    *   CPAN Testers

        The CPAN Testers is a network of smokers who run automated tests on
        uploaded CPAN distributions.

        <http://www.cpantesters.org/distro/D/Data-UUID-Concise>

    *   CPAN Testers Matrix

        The CPAN Testers Matrix is a website that provides a visual overview
        of the test results for a distribution on various Perls/platforms.

        <http://matrix.cpantesters.org/?dist=Data-UUID-Concise>

    *   CPAN Testers Dependencies

        The CPAN Testers Dependencies is a website that shows a chart of the
        test results of all dependencies for a distribution.

        <http://deps.cpantesters.org/?module=Data::UUID::Concise>

  Bugs / Feature Requests
    Please report any bugs or feature requests by email to
    "nrr+bug-DATA-UUID-CONCISE@corvidae.org", or through the web interface
    at <https://github.com/nrr/Data-UUID-Concise/issues>. You will be
    automatically notified of any progress on the request by the system.

  Source Code
    <https://github.com/nrr/Data-UUID-Concise>

      git clone https://github.com/nrr/Data-UUID-Concise.git

AUTHOR
    Nathaniel Reindl <nrr@corvidae.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Nathaniel Reindl.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Data::UUID::Concise

0.121082  2012-04-17 05:13:05 America/Chicago

	- Yet another test release, now with more feeling.

0.121081  2012-04-17 05:12:03 America/Chicago

	- A test release.

0.121080
	Released: 2012-04-17 02:14:42

	- Initial release.


lib/Data/UUID/Concise.pm
use warnings;
use strict;
use strictures;

package Data::UUID::Concise;

use 5.010;
use utf8;
use open qw(:std :utf8);
use charnames qw(:full :short);

use Moo;
use MooX::Types::MooseLike::Base qw(:all);

use Carp;
use Data::UUID;
use Math::BigInt;

use feature qw[ say ];

our $VERSION = '0.121082';    # VERSION

# ABSTRACT: Encode UUIDs to be more concise or communicable
# ENCODING: utf-8

has 'alphabet' => (
    is      => 'rw',
    isa     => Str,
    default => sub {
        '23456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';
    },
);

sub encode
{
    my ( $self, $uuid ) = @_;

    my $output = '';
    my $numeric =
        Math::BigInt->new( ( Data::UUID->new )->to_hexstring( $uuid ) );
    my $alphabet_length = length( $self->alphabet );

    while ( $numeric->is_positive ) {
        my $index = $numeric->copy->bmod( $alphabet_length );
        $output .= substr( $self->alphabet, $index, 1 );
        $numeric->bdiv( $alphabet_length );
    }

    return $output;
}

sub decode
{
    my ( $self, $string ) = @_;

    my $numeric         = Math::BigInt->new;
    my @characters      = split //, $string;
    my $alphabet_length = length( $self->alphabet );

    for my $character ( @characters ) {
        my $value = index $self->alphabet, $character;
        $numeric = $numeric->bmul( $alphabet_length );
        $numeric = $numeric->badd( $value );
    }

    return ( Data::UUID->new )->from_hexstring( $numeric->as_hex );
}

1;

__END__

=pod

=for :stopwords Nathaniel Reindl cpan testmatrix url annocpan anno bugtracker rt cpants
kwalitee diff irc mailto metadata placeholders metacpan

=encoding utf-8

=head1 NAME

Data::UUID::Concise - Encode UUIDs to be more concise or communicable

=head1 VERSION

version 0.121082

=head1 SYNOPSIS

    use Data::UUID::Concise;

    my $duc = Data::UUID::Concise->new();
    my $encoded_uuid = $duc->encode((Data::UUID->new)->create);
    my $decoded_uuid = $duc->decode('M55djt9tt4WoFaL68da9Ef');

=head1 ATTRIBUTES

=head2 alphabet

This is the collection of symbols that are used for the encoding scheme. By
default, a reasonably unambiguous set of characters is used that is reminiscent
of the base 58 scheme used by a rather prominent photo site's URL shortener.

=head1 METHODS

=head2 encode

Encode a Data::UUID instance as a string with the appropriate set of symbols.

=head2 decode

Decode a string with the appropriate set of symbols and return a Data::UUID
instance representing the decoded UUID.

=head1 SUPPORT

=head2 Perldoc

You can find documentation for this module with the perldoc command.

  perldoc Data::UUID::Concise

=head2 Websites

The following websites have more information about this module, and may be of help to you. As always,
in addition to those websites please use your favorite search engine to discover more resources.

=over 4

=item *

Search CPAN

The default CPAN search engine, useful to view POD in HTML format.

L<http://search.cpan.org/dist/Data-UUID-Concise>

=item *

CPAN Ratings

The CPAN Ratings is a website that allows community ratings and reviews of Perl modules.

L<http://cpanratings.perl.org/d/Data-UUID-Concise>

=item *

CPAN Testers

The CPAN Testers is a network of smokers who run automated tests on uploaded CPAN distributions.

L<http://www.cpantesters.org/distro/D/Data-UUID-Concise>

=item *

CPAN Testers Matrix

The CPAN Testers Matrix is a website that provides a visual overview of the test results for a distribution on various Perls/platforms.

L<http://matrix.cpantesters.org/?dist=Data-UUID-Concise>

=item *

CPAN Testers Dependencies

The CPAN Testers Dependencies is a website that shows a chart of the test results of all dependencies for a distribution.

L<http://deps.cpantesters.org/?module=Data::UUID::Concise>

=back

=head2 Bugs / Feature Requests

Please report any bugs or feature requests by email to C<nrr+bug-DATA-UUID-CONCISE@corvidae.org>, or through
the web interface at L<https://github.com/nrr/Data-UUID-Concise/issues>. You will be automatically notified of any
progress on the request by the system.

=head2 Source Code


L<https://github.com/nrr/Data-UUID-Concise>

  git clone https://github.com/nrr/Data-UUID-Concise.git

=head1 AUTHOR

Nathaniel Reindl <nrr@corvidae.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Nathaniel Reindl.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Encode UUIDs to be more concise or communicable",
  "AUTHOR" => "Nathaniel Reindl <nrr\@corvidae.org>",
  "BUILD_REQUIRES" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "Test::Exception" => 0,
    "Test::More" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Data-UUID-Concise",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Data::UUID::Concise",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::UUID" => 0,
    "Math::BigInt" => 0,
    "Moo" => 0,
    "MooX::Types::MooseLike::Base" => 0,
    "charnames" => 0,
    "feature" => 0,
    "open" => 0,
    "strict" => 0,
    "strictures" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.121082",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



